torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,253 @@
+      while (waitpid(-1, 0, WNOHANG) > 0) {}
+    } else if (orig_SIGCHLD_sa.sa_sigaction != (void (*)(int, siginfo_t *, void *)) SIG_DFL) {
+      // SIG_DFL for SIGCHLD is to leave the child as a zombie (do nothing)
+      orig_SIGCHLD_sa.sa_sigaction(sig, info, ctx);
+    }
+  } else {
+    // handler is sa_handler
+    // https://stackoverflow.com/a/24080440
+    if (orig_SIGCHLD_sa.sa_handler == SIG_IGN) {
+      while (waitpid(-1, 0, WNOHANG) > 0) {}
+    } else if (orig_SIGCHLD_sa.sa_handler != SIG_DFL) {
+      orig_SIGCHLD_sa.sa_handler(sig);
+    }
+  }
+}
+
+// returns -1 on error.
+static int isSIGCHLDHanderSet() {
+  struct sigaction sa;
+  int error = sigaction(SIGCHLD, NULL, &sa);
+  if (error == 0) {
+    return ((sa.sa_flags | SA_SIGINFO) != 0) && (sa.sa_sigaction == &handler_SIGCHLD_main);
+  } else {
+    return -1;
+  }
+}
+
+#define SAFECALL(CALL) if ((CALL) == -1) Py_RETURN_FALSE
+
+// We don't want to exit on any SIGCHLD from any child. child_pids is a tuple
+// of pids we are interested in.
+PyObject *THPModule_setMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  HANDLE_TH_ERRORS
+  // assert these types are lock free, just to be safe
+  THPUtils_assert(worker_pids.is_lock_free(), "worker_pids is not lock free");
+  THPUtils_assert(num_worker_pids.is_lock_free(), "num_worker_pids is not lock free");
+
+  THPUtils_assert(PyTuple_Check(child_pids), "_set_main_signal_handlers_for_workers "
+        "expects a tuple, but got %s", THPUtils_typename(child_pids));
+
+  if (comm_pipe[0] == -1) {
+    // we have GIL here so we are fine
+    SAFECALL(pipe(comm_pipe));
+    char c = '_';
+    write(comm_pipe[1], &c, 1);
+  }
+
+  auto size = PyTuple_GET_SIZE(child_pids);
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = PyTuple_GET_ITEM(child_pids, idx);
+    worker_pid_vec.push_back((pid_t) THPUtils_unpackLong(obj));
+  }
+  SAFECALL(updatePIDsArray());
+
+  // To avoid chain calling our handler, check if the current handler is already
+  // set as ours.
+  int set = 0;
+  SAFECALL(set = isSIGCHLDHanderSet());
+  if (set == 0) {
+    SAFECALL(setSignalHandler(SIGCHLD, &handler_SIGCHLD_main, &orig_SIGCHLD_sa));
+  }
+  Py_RETURN_TRUE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPModule_removeMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  HANDLE_TH_ERRORS
+  THPUtils_assert(PyTuple_Check(child_pids), "_remove_main_signal_handlers_for_workers "
+        "expects a tuple, but got %s", THPUtils_typename(child_pids));
+
+  auto size = PyTuple_GET_SIZE(child_pids);
+  std::set<pid_t> pid_set;
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = PyTuple_GET_ITEM(child_pids, idx);
+    pid_set.insert((pid_t) THPUtils_unpackLong(obj));
+  }
+  worker_pid_vec.erase(std::remove_if(worker_pid_vec.begin(), worker_pid_vec.end(),
+     [&](pid_t p){return pid_set.find(p) != pid_set.end();}), worker_pid_vec.end());
+  updatePIDsArray();
```

Comment:
That's why we should use exceptions. You forgot to check the return code

Reply:
Also, since we're going to reallocate the pid pointer anyway, why don't we just keep a `pid_set` instead of that vector?